rfcs/0136-stabilize-incrementally.md

```diff
@@ -0,0 +1,189 @@
+
+The plan below attempts to make all sides happy, despite their seemingly irreconcilable differences.
+The basic thrust is to proceed with stabilization in small steps.
+Two benefits are envisioned:
+
+### Benefit 1: Keep discussions on track.
+
+Threads on Flakes in the past have been impossible to follow because the sheer magnitude of different topics being discussed.
+Small stabilization steps are meant to yield *focused, bounded* discussions that actually read well.
+This also ensures any issues that do arise are actually brought to attention, and not lost within a deluge of other topics.
+
+### Benefit 2: Build trust by starting with the least controversial material.
+
+Flakes are obviously the most controversial, and so they are saved for last.
+The CLI is more broadly popular, but still is a lot of material to discuss.
+The store-only subcommands are Nix's "plumbing" as opposed to "porcelain" commands, and thus have the simplest (if not most user-friendly) purposes.
+This "dusty corner" off the new CLI is rather calm with a very constrained design space, and far less acrimony.
+
+## Conclusion
+
+By starting with these relatively easy material to stabilize, we can prove we can all come together as a community to agree on a design after all.
+This should build trust between opposing individuals and factions, giving us a foundation upon which to tackle the more challenging material in subsequent steps.
+
+since these features became very popular while they are still unstable, there is now both an urgency to stabilize them, and little ability to modify them.
+This is the opposite of how experimental features are supposed to leisurely incubate with both plenty of flexibility to change them, and little rush to stabilize them until they feel ready.
+But, there is little we can do about this at this point, an this RFC recognizes that fact and does *not* try to resist it.
+
+But, we have crossed that Rubicon and there is no turning back; this RFC *doesn't* attempt to change Flakes or the command line 
+
+# Detailed design
+[design]: #detailed-design
+
+Step 1 is technical work, with a self-imposed deadline so we can be sure it doesn't delay stabilization too long.
+The remaining steps are stabilization steps.
+For each of them, a separate RFC or other discussion medium will describe the new interfaces to be stabilized, and solicit feedback.
+
+## Step 1: Split out a store-only Nix, with deadline
+
+This is the same as step one from [RFC 134](https://github.com/NixOS/rfcs/pull/134), upon which this RFC depends.
+
+A deadline is imposed to get that complete, so agreement on this stabilization plan is guaranteed not to add more than a small bounded delay as opposed to stabilizing everything immediately.
+
+## Step 2: Stabilize the store-only Nix CLI
+
+If the deadline for step 1 is met, then we stabilize *just* the command-line interface of the store-only Nix command.
```

Comment:
Actually, even if store-only Nix is not there, can `nix` command be stabilised area by area, probably with store-related commands first?

Reply:
Actually current RFC #134 does not seem to require that eval-part of Nix goes only through the CLI-exposed part of store-Nix, does it?

Reply:
No it does not, but I meant making sure low level commands don't need high level library code.

Reply:
But this doesn't really prove that the store-Nix is useful as plumbing without layering violations (like the current way of specifying a specific output), does it?

Reply:
@7c6f434c This feels like a "necessary vs sufficient" type distinction?

Enforcing a store-only Nix lets us figure out automatically which plumbing commands are virtuously not using higher level layer, but yes it won't tell us if those low level commands are sufficiently expressive. However I think merely having less items to review will allow us in a less automatic but human-driven way find where the gaps are. I would hope to finish my PR for the output selection problem too, for example.

Reply:
If we don't have guarantees of layering either way, a fallback of just manually splitting the commands by semantic area doesn't feel pointless, just a slight reduction from non-perfect level of rigour to a bit lower non-perfect but non-useless level of rigour.

The commitment during the stabilisation is about the interface after all, so going semantic area by semantic area and ensuring consistency there sounds a reasonable approach. 

I am not opposing trying to make store-Nix and ideally later have evaluation-Nix not touching the store but through store-Nix! I am just saying that its failure should not doom step 2, it can be done on a different foundation as well.

Reply:
Oh OK, I thought you were saying we needed a more stringent process, but it sounds like you are saying the opposite. Sure, that makes sense.

Reply:
Sure I agree. I am hoping I get that prev RFC approved / PR done regardless and it's a non-issue, but if I fail then yes it is good that we can still stabilize things incrementally.

Reply:
@7c6f434c It seems you're on the same page now. Maybe you can mark the conversation as resolved? That would make it easier to see if there's still something left to discuss or not.